### PR TITLE
[GAL-2124] Invalidate cache on fetch errors

### PR DESCRIPTION
**Pull request by jotadrilo**
Signed-off-by: Joseda Rios <jdrios@vmware.com>

We detected problems during the charts indexing process (see #167).

However, while I could not detect the conditions under which this process fails to generate the corrupted tarball, I spotted that we are not closing file descriptors (nor HTTP body responses) correctly.

This patch refactors the OCI and Helm repos to reuse the same logic to fetch and cache charts using the HTTP protocol, and to correctly close file descriptors and responses. It also removes the cached chart in case of an error writing/closing the file. This will avoid unexpected issues during the next sync.

Additionally, it moves the disk cache implementation to a subpackage to decouple the interface from the implementation.

New runs of the sync process haven't generated corrupted tarballs, but I cannot ensure this patch is the right bug fix.

### CODE REVIEW

File: internal/cache/cachedisk/cachedisk.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	klog.V(4).Infof("cache hit { op:has, id:%s, filename:%s }", c.id, filename)
+	return ok
+}
+
+// Read reads a cached file named by filename and writes the contents to the
+// provider w writer.
+//
+// If the file does not exist, Read returns an error that satisfies
+// errors.IsNotFound().
+func (c *Cache) Read(w io.Writer, filename string) error {
```
Comment (tompizmor) [inline internal/cache/cachedisk/cachedisk.go:64]: I don't see where you are using this method...

File: internal/cache/cachedisk/cachedisk.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	}
+	klog.V(4).Infof("cache hit { op:read, id:%s, filename:%s }", c.id, filename)
+	return nil
+}
+
+// Store stores the provided bytes in a cache file named by filename
+//
+// If the file exists, Store returns an error that satisfies
+// errors.IsAlreadyExists(). Otherwise, Store creates it with permissions perm
+func (c *Cache) Store(r io.Reader, filename string) error {
```
Comment (tompizmor) [inline internal/cache/cachedisk/cachedisk.go:83]: same...is this being used?

File: internal/cache/cachedisk/cachedisk.go (hunk 1 above)
Comment (jotadrilo) [reply internal/cache/cachedisk/cachedisk.go:64]: I am not using this in the code that I have modified. Note that this file was moved from `cache/store.go`

File: internal/cache/cachedisk/cachedisk.go (hunk 2 above)
Comment (jotadrilo) [reply internal/cache/cachedisk/cachedisk.go:83]: same

**Pull request merged**